src/lib/snapp.ts

```diff
@@ -202,19 +211,63 @@ export function state<A>(ty: AsFieldElements<A>) {
  * }
  * ```
  */
-export function method(
-  target: any,
-  propertyName: string,
+export function method<T extends SmartContract>(
+  target: T & { constructor: any },
+  methodName: keyof T & string,
   _descriptor?: PropertyDescriptor
-): any {}
+) {
+  const SnappClass = target.constructor;
+  if (reservedPropNames.has(methodName)) {
+    throw Error(`Property name ${methodName} is reserved.`);
+  }
+  if (typeof target[methodName] !== 'function') {
+    throw Error(
+      `@method decorator was applied to ${methodName} which is not a function.`
+    );
+  }
+  let paramTypes = Reflect.getMetadata('design:paramtypes', target, methodName);
+  let witnessArgs = [];
+  let proofArgs = [];
+  let args = [];
+  for (let i = 0; i < paramTypes.length; i++) {
+    let Parameter = paramTypes[i];
+    if (isProof(Parameter)) {
+      args.push({ type: 'proof', index: proofArgs.length });
+      proofArgs.push(Parameter);
+    } else if (isAsFields(Parameter)) {
+      args.push({ type: 'witness', index: witnessArgs.length });
+      witnessArgs.push(Parameter);
+    } else {
+      throw Error(
+        `Argument ${i} of method ${methodName} is not a valid circuit value.`
```

Comment:
(not in this PR, I made #130 to track it). 
We should include details about why this is a not a valid circuit value and what a circuit value is in this error.